### PR TITLE
Solo mini-game

### LINKED ISSUE #856: One player game
[Oneplayer.zip](https://github.com/virtualcommons/port-of-mars/files/11267120/Oneplayer.zip)

In the Zipfile you find a word document explaining the one player game and a Netlogo file where we simulated outcomes with a player who did pretty much making random decisions (I use a zip file since nlogo files are not recognized by Github). I did not include the academic reasoning why we do certain treatments in the word document.

### LINKED ISSUE #882: Track and display a highscores leaderboard for solo game mode
New tab/section in leaderboard and personal game history for indexing highscores on the solo game (#856)

Use high-scores (max), as opposed to the cumulative points used in the current leaderboard ranking

### LINKED ISSUE #888: integrate "Events this round" and "All events" widgets
Consider integrating the "Events This Round" display into the "All Events" widget

- Rename to "Events"
- visually highlight the events received this round in the "Events" widget and scroll to them automatically (if full event deck is being shown or there is a history of past events)

### PR DESCRIPTION
**Pull request by sgfost**
Main development branch which adds the implementation of a solo game mode as described in #856

---
- [x] resolves #856
- [x] resolves #882 
- [x] resolves #888

### CODE REVIEW

File: client/src/components/sologame/SegmentedBar.vue [hunk 1]
```diff
@@ -147,6 +139,8 @@ export default class SegmentedBar extends Vue {
   }
 
   mounted() {
+    // FIXME: consider if we really want to use scroll wheel to increment / decrement the value, keyboard input might be
+    // better for accessibility (+ / -) (<- ->)?
```
Comment (sgfost) [inline client/src/components/sologame/SegmentedBar.vue:143]: keyboard controls are a good idea, L/R arrow keys and enter for investing/continuing through events

the scroll thing only happens when the cursor is over the input bar so it hopefully wouldn't interfere with anything too much, though probably unnecessary if we have keyboard controls

File: server/src/services/sologame.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    for (const card of cards) {
+      const drawAmt = getRandomIntInclusive(card.drawMin, card.drawMax);
+      for (let i = 0; i < drawAmt; i++) {
+        // card effects are encoded in the db with a range of possible rolls and
+        // a multiplier for each value (sys health, points, .. ), this is typically
+        // 0, 1 or -1
+        const roll = getRandomIntInclusive(card.rollMin, card.rollMax);
+        // fill out templates from the db with the actual roll value and pluralization
+        const effectText = card.effect
+          .replace("{roll}", roll.toString())
```
Comment (alee) [inline server/src/services/sologame.ts:34]: these dynamics template strings could also be set up as functions in the EventCards themselves, something like https://stackoverflow.com/a/40147828/93370

(for the future)

File: server/src/services/stats.ts [hunk 3]
```diff
@@ -101,4 +107,71 @@ export class StatsService extends BaseService {
...
+      highscore = newHighscore;
+    }
+  }
+
+  async getSoloHighScoreData(limit: number): Promise<SoloHighScoreData> {
+    // get the top players in the solo highscores table
+    const highscoresData = await this.em
+      .getRepository(SoloHighScore)
+      .createQueryBuilder("highscores")
+      .select("ROW_NUMBER() OVER (ORDER BY highscores.pointsPerRound DESC)::integer as rank")
```
Comment (alee) [inline server/src/services/stats.ts:147]: this SQL query would benefit from more detailed comments @virtualcommons/port-of-mars-dev

**Review by alee — APPROVED** [review]
great work @virtualcommons/port-of-mars-dev team! This looks really solid. I had a few issues with overall codebase consistency but these are mostly simple changes like making imports consistent etc. 

Changes to the schema are a thornier issue but we can discuss this in this afternoon's meeting

**Comment by sgfost** [comment]
@all-contributors please add @sabrinanel3 for code, docs, design

@alee just remembered @sabrinanel3 is missing from the @.virtualcommons/port-of-mars-dev team as well and I can't seem to add people to that

**Comment by sgfost** [comment]
@all-contributors please add @sabrinanel3 for code, docs, design

**Pull request merged**